### ISSUE TITLE
Notification delivered into device 2 or 3 times (duplicated notifications)

### ISSUE THREAD

**Issue by jsifalda** [node-apn/node-apn#558]
Hey guys, does anybody else experience with the same issue? Sometimes push notifications is duplicated two times or even three times. I can confirm that the push notification has been sent just just once from our system.

This is how the actual sending looks like:

```js
var sendNotification = (env, notification, token, shouldBeLogged) => {
  if (token && !Array.isArray(token)) {
    token = [token]
  }

  return new Promise((resolve, reject) => {
    var t = (notification && notification.payload && notification.payload.body) ? notification.payload.body : ''
    console.log('iOS: Pushing notification to device', token, t)

    env
    .send(notification, token)
    .then((response) => {
      var sent = response.sent || []
      var failed = response.failed || []
      ....
     })

     resolve()
  })
}
....
```
Env is actual connection which is created at the boot of microservice and maintained since that. We have 2 connections created - one is for prod and second for dev push notifications.

And generally, how this sending works... it is microservice dedicated to sending of push notifications. Push notifications are distributed via rabbit and collected in this microservice. Only one task can be processing at the same time. However, one task can be sending one push notification to many users (that why "resolve" is not in "then" callback). Each user has got different notification object because different badge count etc. 

It looks like problem hidden somewhere in version of this lib or on apple side.

Any ideas? Would be highly appreciated.
Jiri

PS. Version of apn lib is 2.1.0

**Comment by jplanders** [comment]
I'm not sure if this is what you are running into but I'll share it with you just in case.

Note: This happened to me using a different library.

In testing I would try to send a notification to 3 devices but the first device would get the notification 3 times.  I was iterating through an array of device tokens using async.each.  Ultimately, I changed async.each for async.eachLimit only allowing for 1 async operation at a time and that resolved the issue I was having.

**Comment by jsifalda** [comment]
Well, it is not really solution...

**Comment by florianreinhart** [comment]
@jsifalda Is this still an issue for you?

I've never seen this. Could it be an error in your code?

**Comment by jsifalda** [comment]
Yes, we have not figured out what is causing it... probably apple's bug because it common for other apps to see duplicated push notifications as well... especially for those who push a lot of pns (eg. live sports apps)
